http/util.go

```diff
@@ -36,26 +37,46 @@ func rateLimitQuotaWrapping(handler http.Handler, core *vault.Core) http.Handler
 			respondError(w, http.StatusInternalServerError, err)
 			return
 		}
-		req := w.(*LogicalResponseWriter).request
+
+		path, op, status, err := buildLogicalPathAndOp(r)
+		if err != nil || status != 0 {
+			respondError(w, status, err)
+			return
+		}
+
 		quotaResp, err := core.ApplyRateLimitQuota(&quotas.Request{
 			Type:          quotas.TypeRateLimit,
-			Path:          req.Path,
-			MountPath:     strings.TrimPrefix(req.MountPoint, ns.Path),
+			Path:          path,
+			MountPath:     strings.TrimPrefix(core.MatchingMount(r.Context(), path), ns.Path),
 			NamespacePath: ns.Path,
 			ClientAddress: parseRemoteIPAddress(r),
 		})
 		if err != nil {
-			core.Logger().Error("failed to apply quota", "path", req.Path, "error", err)
+			core.Logger().Error("failed to apply quota", "path", path, "error", err)
 			respondError(w, http.StatusUnprocessableEntity, err)
 			return
 		}
 
 		if !quotaResp.Allowed {
-			quotaErr := errwrap.Wrapf(fmt.Sprintf("request path %q: {{err}}", req.Path), quotas.ErrRateLimitQuotaExceeded)
+			quotaErr := errwrap.Wrapf(fmt.Sprintf("request path %q: {{err}}", path), quotas.ErrRateLimitQuotaExceeded)
 			respondError(w, http.StatusTooManyRequests, quotaErr)
 
 			if core.Logger().IsTrace() {
-				core.Logger().Trace("request rejected due to lease count quota violation", "request_path", req.Path)
+				core.Logger().Trace("request rejected due to lease count quota violation", "request_path", path)
+			}
+
+			requestId, err := uuid.GenerateUUID()
+			if err != nil {
+				respondError(w, http.StatusBadRequest, errwrap.Wrapf("failed to generate identifier for the request: {{err}}", err))
+				return
+			}
+
+			req := &logical.Request{
+				ID:         requestId,
+				Operation:  op,
+				Path:       path,
+				Connection: getConnection(r),
+				Headers:    r.Header,
 			}
 
 			if core.RateLimitAuditLoggingEnabled() {
```

Comment:
Since we're always returning within here, we could call `buildLogicalRequestNoAuth` to have it build us the logical request for auditing instead of manually instantiating one and possibly leaving out fields (noticed that `Data` was left out).

Reply:
Done.